Sources/Storage/Disk/Handler/DiskPersistentHandler.swift

```diff
@@ -0,0 +1,120 @@
+//
+//  DiskManager.swift
+//  StatsdClient
+//
+//  Created by Nghia Tran on 10/8/17.
+//  Copyright © 2017 StatsdClient. All rights reserved.
+//
+
+import Foundation
+
+class DiskPersistentHandler: PersistentHandler {
+
+    // MARK: - Variable
+    fileprivate let handler: FileManager
+    fileprivate let path: String
+    fileprivate let defaultEnumerator: [URLResourceKey] = [
+        .isDirectoryKey,
+        .contentModificationDateKey,
+        .totalFileAllocatedSizeKey,
+    ]
+
+    // Config
+    public let config: DiskConfigurable
+
+    // MARK: - Init
+    init(config: DiskConfigurable, hanlder: FileManager = FileManager.default) throws {
+        self.config = config
+        self.handler = hanlder
+
+        // Create path
+        guard let pathFolder = config.pathFolder else {
+            throw DiskError.invalidCacheFolder
+        }
+
+        path = pathFolder
+
+        // Try to create folder
+        try createCacheFolder()
+    }
+
+    // MARK: - Public
+    func makeFilePath(_ key: String) -> String {
+        return "\(path)/\(key.toBase64())"
+    }
+
+    func write<T: Codable>(_ item: T, key: String, attribute: [FileAttributeKey: Any]?) throws {
+        let pathFile = makeFilePath(key)
+        let data = try JSONEncoder().encode(item)
+        handler.createFile(atPath: pathFile, contents: data, attributes: attribute)
+    }
+
+    func get<T: Codable>(key: String, type: T.Type) throws -> T {
+        let pathFile = makeFilePath(key)
+        let data = try Data(contentsOf: pathFile.fileURL(), options: .alwaysMapped)
+        return try JSONDecoder().decode(T.self, from: data)
+    }
+
+    func getAll<T: Codable>(type: T.Type) throws -> [T] {
+
+        // Get all URLs in directory folder
+        let fileURLs = allFileURLs()
+
+        // Map to T
+        let decoder = JSONDecoder()
+        return try fileURLs.map { (url) -> T in
```

Comment:
Hmm, this is not ideal, not only the performance is terrible, but if there is one corrupted file, the whole dataset is garbage. 

Maybe `getAll` should not throws error, but instead try to get everything it can. Something like

```
return fileURLs.flatMap {
            guard let data = try? Data(contentsOf: $0, options: .alwaysMapped) else {
                return nil
            }
            return try? JSONDecoder().decode(T.self, from: data)
        }
```

Reply:
That's a great idea. `flatMap` has the ability to ignore `nil` 👍